### PR TITLE
Allow server admin to get admin bit in rooms where local user is an admin

### PR DESCRIPTION
**Pull request by erikjohnston**
This adds an admin API that allows a server admin to get power in a room if a local user has power in a room. Will also invite the user if they're not in the room and its a private room. Can specify another user (rather than the admin user) to be granted power, e.g.:

```
POST /_synapse/admin/v1/make_room_admin/<room_id_or_alias>
{
    "user_id": "@foo:example.com"
}
```

### CODE REVIEW

File: changelog.d/8756.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Add admin API that let's server admins get power in rooms that local users have power in.
```
Comment (ara4n) [inline changelog.d/8756.feature:1]:
```suggestion
Add admin API that lets server admins get power in rooms that local users have power in.
```

File: synapse/rest/admin/rooms.py [hunk 2]
```diff
@@ -338,3 +341,97 @@ async def on_POST(self, request, room_identifier):
         )
 
         return 200, {"room_id": room_id}
+
+
+class MakeRoomAdminRoomServlet(RestServlet):
```
Comment (ara4n) [inline synapse/rest/admin/rooms.py:346]: can we have a matching /docs/admin entry please?

File: synapse/rest/admin/rooms.py [hunk 3]
```diff
@@ -338,3 +341,97 @@ async def on_POST(self, request, room_identifier):
...
+
+        if power_levels is not None:
+            admin_users = [
+                user_id
+                for user_id, power in power_levels.content.get("users")
+                if self.is_mine_id(user_id)
+            ]
+            admin_users.sort(key=lambda user: power_levels.content.get("users")[user])
+
+            admin_user_id = admin_users[-1]
```
Comment (richvdh) [inline synapse/rest/admin/rooms.py:391]: what if there are no local users with a high enough PL to do this operation?

File: synapse/rest/admin/rooms.py [hunk 4]
```diff
@@ -338,3 +341,97 @@ async def on_POST(self, request, room_identifier):
...
+
+            pl_content = power_levels.content
```
Comment (auscompgeek) [inline synapse/rest/admin/rooms.py:393]: Is `pl_content` defined anywhere else? I can't see where it would be defined if we end up in the else block below instead.

Comment (clokep) [reply synapse/rest/admin/rooms.py:413]: Looks like 628e84dd8d77cd1a792db9e010c670a345244cbe fixes this.

File: synapse/rest/admin/rooms.py (hunk 3 above)
Comment (clokep) [reply synapse/rest/admin/rooms.py:411]: @erikjohnston I think this is an outstanding comment?

File: synapse/rest/admin/rooms.py [hunk 5]
```diff
@@ -340,3 +343,136 @@ async def on_POST(self, request, room_identifier):
         )
 
         return 200, {"room_id": room_id}
+
+
+class MakeRoomAdminRoomServlet(RestServlet):
```
Comment (clokep) [inline synapse/rest/admin/rooms.py:348]: Is this  meant to be named `MakeRoomAdminRestServlet`?

File: synapse/rest/admin/rooms.py [hunk 6]
```diff
@@ -340,3 +343,136 @@ async def on_POST(self, request, room_identifier):
...
+    """Allows a server admin to get power in a room if a local user has power in
+    a room. Will also invite the user if they're not in the room and its a
```
Comment (clokep) [inline synapse/rest/admin/rooms.py:350]:
```suggestion
    a room. Will also invite the user if they're not in the room and it's a
```

File: synapse/rest/admin/rooms.py [hunk 7]
```diff
@@ -340,3 +343,136 @@ async def on_POST(self, request, room_identifier):
...
+            room_alias = RoomAlias.from_string(room_identifier)
+            room_id, _ = await self.room_member_handler.lookup_room_alias(room_alias)
+            room_id = room_id.to_string()
+        else:
+            raise SynapseError(
+                400, "%s was not legal room ID or room alias" % (room_identifier,)
+            )
+
+        # Which user to grant room admin rights to.
+        user_to_add = content.get("user_id", requester.user.to_string())
```
Comment (clokep) [inline synapse/rest/admin/rooms.py:388]: Would it make sense to ensure this is a local user? Maybe it doesn't matter.

File: synapse/rest/admin/rooms.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/rest/admin/rooms.py:411]: D'oh, forgot to respond.

This has also changed. We try and change the power levels with the local user with the maximal power level, and if that fails with an `AuthError` we convert it to a 400 (and there is even a test for this)

File: synapse/rest/admin/rooms.py [hunk 8]
```diff
@@ -340,3 +343,136 @@ async def on_POST(self, request, room_identifier):
...
+
+        create_event = room_state[(EventTypes.Create, "")]
+        power_levels = room_state.get((EventTypes.PowerLevels, ""))
+
+        if power_levels is not None:
+            # We pick the local user with the highest power.
+            user_power = power_levels.content.get("users", {})
+            admin_users = [
+                user_id
+                for user_id, power in user_power.items()
```
Comment (clokep) [inline synapse/rest/admin/rooms.py:403]:
```suggestion
                for user_id in user_power.keys()
```

File: synapse/rest/admin/rooms.py [hunk 9]
```diff
@@ -340,3 +343,136 @@ async def on_POST(self, request, room_identifier):
...
+                },
+            )
+        except AuthError:
+            # The admin user we found turned out not to have enough power.
+            raise SynapseError(
+                400, "No local admin user in room with power to update power levels."
+            )
+
+        # Now we check if the user we're granting admin rights to is already in
+        # the room. If not and its not a public room we invite them.
```
Comment (clokep) [inline synapse/rest/admin/rooms.py:451]:
```suggestion
        # the room. If not and it's not a public room we invite them.
```

File: tests/rest/admin/test_room.py [hunk 10]
```diff
@@ -1437,6 +1438,138 @@ def test_join_private_room_if_owner(self):
...
+            self.creator, tok=self.creator_tok, is_public=True
+        )
+
+        request, channel = self.make_request(
+            "POST",
+            "/_synapse/admin/v1/make_room_admin/{}".format(room_id),
+            content={},
+            access_token=self.admin_user_tok,
+        )
+        self.render(request)
```
Comment (clokep) [inline tests/rest/admin/test_room.py:1476]: I think you need to merge in `develop` since this `render` call is now unnecessary.

File: tests/rest/admin/test_room.py [hunk 11]
```diff
@@ -1437,6 +1438,138 @@ def test_join_private_room_if_owner(self):
...
+
+        self.assertEqual(200, int(channel.result["code"]), msg=channel.result["body"])
+
+        # Now we test that we can join the room (we should have received an
+        # inviate) and can ban a user.
```
Comment (clokep) [inline tests/rest/admin/test_room.py:1508]:
```suggestion
        # invite) and can ban a user.
```

File: docs/admin_api/rooms.md [hunk 12]
```diff
@@ -451,3 +451,19 @@ The following fields are returned in the JSON response body:
 * `local_aliases` - An array of strings representing the local aliases that were migrated from
                     the old room to the new.
 * `new_room_id` - A string representing the room ID of the new room.
+
+
+# Make Room Admin API
+
+Grants the server admin power in a room if a local user has power in the room;
+inviting the server admin if they're not in the room and its not a publically
```
Comment (clokep) [inline docs/admin_api/rooms.md:459]:
```suggestion
inviting the server admin if they're not in the room and it's not a publically
```

File: docs/admin_api/rooms.md [hunk 13]
```diff
@@ -451,3 +451,19 @@ The following fields are returned in the JSON response body:
...
+joinable room.
```
Comment (clokep) [inline docs/admin_api/rooms.md:460]: I find this a bit hard to parse, maybe the following is clearer:

> Grants another user the highest power available to a local user who is in the room.
> If the user is not in the room, and it is not publicly joinable, then invite the user.
>
> By default the server admin (the caller) is granted power, but another user can
> optionally be specified, e.g.:

**Review by clokep — CHANGES_REQUESTED** [review]
Looks reasonable overall. I left a few minor comments and there's one previous outstanding comment.

**Comment by awesome-manuel** [comment]
> ```
> POST /_synapse/admin/v1/make_room_admin/<room_id_or_alias>
> ```

I think it's better to organize endpoints in a logical tree. So my suggestion is to use
> ```
> POST /_synapse/admin/v1/rooms/<room_id_or_alias>/make_room_admin
> ```

**Comment by clokep** [comment]
> > ```
> > POST /_synapse/admin/v1/make_room_admin/<room_id_or_alias>
> > ```
> 
> I think it's better to organize endpoints in a logical tree. So my suggestion is to use
> 
> ```
> POST /_synapse/admin/v1/rooms/<room_id_or_alias>/make_room_admin
> ```

I agree. I think some of our current admin API paths could be improved to match this style! We would need to keep the old API path around, but this is easy enough. That's out of scope for this PR, but would appreciate someone making an issue or PR fixing it.

File: synapse/rest/admin/rooms.py (hunk 7 above)
Comment (erikjohnston) [reply synapse/rest/admin/rooms.py:413]: I think it's fine to set it as a remote user, a bit weird but :shrug:

**Comment by erikjohnston** [comment]
I've changed the API to be under `/rooms/`

**Comment by erikjohnston** [comment]
Sorry for the ridiculous delay in coming back to this PR.... but at least I got back to it before Christmas! :christmas_tree:

**Review by clokep — APPROVED** [review]
